src/QsCompiler/CompilerOptimizations/StatementRemoving.fs

```diff
@@ -0,0 +1,129 @@
+module Microsoft.Quantum.QsCompiler.CompilerOptimization.StatementRemoving
+
+open System
+open System.Collections.Immutable
+open Microsoft.Quantum.QsCompiler.SyntaxExtensions
+open Microsoft.Quantum.QsCompiler.SyntaxTree
+open Microsoft.Quantum.QsCompiler.Transformations.Core
+
+open ComputationExpressions
+open Utils
+open OptimizingTransformation
+open VariableRenaming
+
+
+/// Returns whether all variables in a symbol tuple are discarded
+let rec private isAllDiscarded = function
+| DiscardedItem -> true
+| VariableNameTuple items -> Seq.forall isAllDiscarded items
+| _ -> false
+
+/// Returns whether a statements is purely classical.
+/// The statement must have no classical or quantum side effects other than defining a variable.
+let private isPureClassical = function
+| QsVariableDeclaration s when not (s.Rhs.InferredInformation.HasLocalQuantumDependency) -> true
+| _ -> false
+
+/// Returns whether a statement is purely quantum.
+/// The statement must have no classical side effects, but can have quantum side effects.
+let private isPureQuantum = function
+| QsVariableDeclaration s when isAllDiscarded s.Lhs -> true
+| QsExpressionStatement _ -> true
+| _ -> false
+
+/// Reorders a list of statements such that the pure classical statements occur before the pure quantum statements
+let rec private reorderStatements (stmtList: QsStatement list): QsStatement list =
+    match stmtList with
+    | [] -> []
+    | head :: tail ->
+        let tail = reorderStatements tail
+        match head :: tail with
+        | head1 :: head2 :: tail
+            when isPureQuantum head1.Statement && isPureClassical head2.Statement ->
+            head2 :: reorderStatements (head1 :: tail)
+        | a -> a
+
+
+/// The SyntaxTreeTransformation used to remove useless statements
```

Comment:
Looks like this is doing a bit more than just removing unnecessary statements. Consider whether splitting the functionality into individual transformations makes sense, and update the doc comment in any case.

Reply:
Yes, splitting would be good.